lib/Transforms/Scalar/HoistConstantArray.cpp

```diff
@@ -0,0 +1,376 @@
+  for (User *U : m_Alloca->users()) {
+    GEPOperator *gep = dyn_cast<GEPOperator>(U);
+    if (!gep)
+      return false;
+
+    geps.push_back(gep);
+  }
+
+  return true;
+}
+
+// Check that all gep uses are valid.
+// A valid use is either
+//  1. A store of a constant value that does not overwrite an existing constant
+//     with a different value.
+//  2. A load instruction.
+// Any other use is considered invalid.
+bool CandidateArray::AllGEPUsersAreValid(GEPOperator *gep) {
+  for (User *U : gep->users()) {
+    if (StoreInst *SI = dyn_cast<StoreInst>(U)) {
+      if (!AnalyzeStore(SI))
+        return false;
+    }
+    else if (!isa<LoadInst>(U)) {
+      return false;
+    }
+  }
+
+  return true;
+}
+
+// Analyze all uses of the array to see if it qualifes as a constant array.
+// We check the following conditions:
+//  1. Make sure alloca is only used by GEP.
+//  2. Make sure GEP is only used in load/store.
+//  3. Make sure all stores have constant indicies.
+//  4. Make sure all stores are constants.
+//  5. Make sure all stores to same location are the same constant.
+void CandidateArray::AnalyzeUses() {
+  m_IsConstArray = false;
+  std::vector<GEPOperator *> geps;
+  if (!AllArrayUsersAreGEP(geps))
+    return;
+
+  for (GEPOperator *gep : geps)
+    if (!AllGEPUsersAreValid(gep))
+      return;
+
+  m_IsConstArray = true;
+}
+
+// Analyze a store to see if it is a valid constant store.
+// A valid store will write a constant value to a known (constant) location.
+bool CandidateArray::AnalyzeStore(StoreInst *SI) {
+  if (!isa<Constant>(SI->getValueOperand()))
+    return false;
+
+  GEPOperator *gep = cast<GEPOperator>(SI->getPointerOperand());
+  if (!gep->hasAllConstantIndices())
+    return false;
+
+  assert(gep->getPointerOperand() == m_Alloca);
+  assert(gep->getNumIndices() == 2);
+
+  ConstantInt *ptrOffset = cast<ConstantInt>(gep->getOperand(1));
+  ConstantInt *index = cast<ConstantInt>(gep->getOperand(2));
+
+  // Non-zero offset is unexpected, but could occur in the wild. Bail out if we see it.
+  if (!ptrOffset->isZero())
+    return false;
+
+  return StoreConstant(index->getLimitedValue(), cast<Constant>(SI->getValueOperand()));
+}
+
+// Check if the store is valid and record the value if so.
+// A valid constant store is either:
+//  1. A store of a new constant
+//  2. A store of the same constant to the same location
+bool CandidateArray::StoreConstant(uint64_t index, Constant *value) {
+  EnsureSize();
+  size_t i = static_cast<size_t>(index);
+  if (index >= m_Values.size())
+    return false;
+  if (m_Values[i] == UndefElement())
+    m_Values[i] = value;
+
+  return m_Values[i] == value;
+}
+
+// We lazily create the values array until we have a store of a
+// constant that we need to remember. This avoids memory overhead
+// for obviously non-constant arrays.
+void CandidateArray::EnsureSize() {
+  uint64_t arraySize = m_ArrayType->getNumElements();
+  if (m_Values.size() != arraySize) {
```

Comment:
Can simply check against zero first, then move the ->getNumElements() inside the branch (after all, it gets initialized at-most-one to the value).